### PR TITLE
Add YOLOv3 loss operator for YOLOv3 model

### PR DESCRIPTION
**Pull request by heavengate**

### CODE REVIEW

File: paddle/fluid/operators/yolov3_loss_op.cc [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+#include "paddle/fluid/operators/yolov3_loss_op.h"
+#include "paddle/fluid/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+using framework::Tensor;
+
+class Yolov3LossOp : public framework::OperatorWithKernel {
```
Comment (qingqing01) [inline paddle/fluid/operators/yolov3_loss_op.cc:20]: 在DarkNet里这个loss名字叫啥？ 这个loss可以用于yolo v2吗？

File: paddle/fluid/operators/yolov3_loss_op.cc [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+    std::vector<int64_t> dim_out({1});
+    ctx->SetOutputDim("Loss", framework::make_ddim(dim_out));
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
+    return framework::OpKernelType(
+        framework::ToDataType(ctx.Input<Tensor>("X")->type()), ctx.GetPlace());
```
Comment (qingqing01) [inline paddle/fluid/operators/yolov3_loss_op.cc]: .h里实现看着是不支持GPU的，如果是，这里`ctx.GetPlace()` -> `CPUPlace()`

File: paddle/fluid/operators/yolov3_loss_op.cc [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  }
+};
+
+class Yolov3LossOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X",
+             "The input tensor of bilinear interpolation, "
+             "This is a 4-D tensor with shape of [N, C, H, W]");
```
Comment (qingqing01) [inline paddle/fluid/operators/yolov3_loss_op.cc]: 注释不对，需修改

File: paddle/fluid/operators/yolov3_loss_op.cu [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+#include "paddle/fluid/operators/yolov3_loss_op.h"
+#include "paddle/fluid/platform/cuda_primitives.h"
+
+namespace ops = paddle::operators;
+REGISTER_OP_CUDA_KERNEL(
+    yolov3_loss,
+    ops::Yolov3LossKernel<paddle::platform::CUDADeviceContext, float>);
+REGISTER_OP_CUDA_KERNEL(
+    yolov3_loss_grad,
+    ops::Yolov3LossGradKernel<paddle::platform::CUDADeviceContext, float>);
```
Comment (qingqing01) [inline paddle/fluid/operators/yolov3_loss_op.cu]: 同上，如果.h里的实现不支持GPU，不用添加.cu文件

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -8243,6 +8244,74 @@ def log_loss(input, label, epsilon=1e-4, name=None):
     return loss
 
 
+@templatedoc(op_type="yolov3_loss")
+def yolov3_loss(x, gtbox, anchors, class_num, ignore_thresh, name=None):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: 放在detection.py里吧

File: paddle/fluid/operators/yolov3_loss_op.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/operators/yolov3_loss_op.cc:20]: Darknet里面是以layer为粒度的，yolov2最后一层叫region层，yolov3叫yolo层，这个loss和yolov2的loss有差别，不可以用于yolov2

File: paddle/fluid/operators/yolov3_loss_op.cc (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/yolov3_loss_op.cc]: Done

File: paddle/fluid/operators/yolov3_loss_op.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/yolov3_loss_op.cc]: Done

File: paddle/fluid/operators/yolov3_loss_op.cu (hunk 4 above)
Comment (heavengate) [reply paddle/fluid/operators/yolov3_loss_op.cu]: Done

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done

**Review by heavengate — COMMENTED** [review]
Thank for reviewing and all have been fixed

**Review by qingqing01 — DISMISSED** [review]

File: python/paddle/fluid/layers/detection.py [hunk 6]
```diff
@@ -404,6 +406,102 @@ def polygon_box_transform(input, name=None):
...
+                lambda_conf_obj=None,
+                lambda_conf_noobj=None,
+                lambda_class=None,
+                name=None):
+    """
+    ${comment}
+
+    Args:
+        x (Variable): ${x_comment}
+        gtbox (Variable): groud truth boxes, shoulb be in shape of [N, B, 5],
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: shoulb -> should

File: python/paddle/fluid/layers/detection.py [hunk 7]
```diff
@@ -404,6 +406,102 @@ def polygon_box_transform(input, name=None):
...
+                          in the third dimenstion, class_id, x, y, w, h should
+                          be stored and x, y, w, h should be relative valud of
+                          input image.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: In SSD and Faster-RCNN, the class_id and gt_box are two variables. So, maybe also two variables here ?

File: python/paddle/fluid/layers/detection.py [hunk 8]
```diff
@@ -404,6 +406,102 @@ def polygon_box_transform(input, name=None):
...
+def yolov3_loss(x,
+                gtbox,
+                anchors,
+                class_num,
+                ignore_thresh,
+                lambda_xy=None,
+                lambda_wh=None,
+                lambda_conf_obj=None,
+                lambda_conf_noobj=None,
+                lambda_class=None,
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: `lambda_` 前缀是不是不够直观？ weight or coefficient？DarkNet里叫啥？

File: python/paddle/fluid/tests/unittests/test_layers.py [hunk 9]
```diff
@@ -911,6 +911,15 @@ def test_affine_grid(self):
...
         print(str(program))
 
+    def test_yolov3_loss(self):
+        program = Program()
+        with program_guard(program):
+            x = layers.data(name='x', shape=[30, 7, 7], dtype='float32')
+            gtbox = layers.data(name='gtbox', shape=[10, 5], dtype='float32')
+            loss = layers.yolov3_loss(x, gtbox, [10, 13, 30, 13], 10, 0.5)
+
+            self.assertIsNotNone(loss)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_layers.py]: 单测也可以放在test_detection.py里

File: python/paddle/fluid/layers/detection.py [hunk 10]
```diff
@@ -404,6 +406,102 @@ def polygon_box_transform(input, name=None):
...
+@templatedoc(op_type="yolov3_loss")
+def yolov3_loss(x,
+                gtbox,
+                anchors,
+                class_num,
+                ignore_thresh,
+                lambda_xy=None,
+                lambda_wh=None,
+                lambda_conf_obj=None,
+                lambda_conf_noobj=None,
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: obj -> target ?

Comment (heavengate) [reply python/paddle/fluid/layers/detection.py]: Done

File: python/paddle/fluid/layers/detection.py (hunk 8 above)
Comment (heavengate) [reply python/paddle/fluid/layers/detection.py]: Done. 
Darknet 里面没有实现权重，yolov3的paper也比较简略，我看有人keras和pytorch实现的版本带了权重，应该是沿用yolov2的 \lambda 权重系数实现的，觉得模型调优时可能用得上就实现了下

File: python/paddle/fluid/layers/detection.py (hunk 7 above)
Comment (heavengate) [reply python/paddle/fluid/layers/detection.py]: Done

File: python/paddle/fluid/tests/unittests/test_layers.py (hunk 9 above)
Comment (heavengate) [reply python/paddle/fluid/tests/unittests/test_layers.py]: Done

**Review by heavengate — COMMENTED** [review]
Thanks for your suggestions and all have been fixed.

**Review by qingqing01 — DISMISSED** [review]

File: paddle/fluid/API.spec [hunk 11]
```diff
@@ -288,6 +288,7 @@ paddle.fluid.layers.generate_proposals ArgSpec(args=['scores', 'bbox_deltas', 'i
 paddle.fluid.layers.iou_similarity ArgSpec(args=['x', 'y', 'name'], varargs=None, keywords=None, defaults=(None,))
 paddle.fluid.layers.box_coder ArgSpec(args=['prior_box', 'prior_box_var', 'target_box', 'code_type', 'box_normalized', 'name'], varargs=None, keywords=None, defaults=('encode_center_size', True, None))
 paddle.fluid.layers.polygon_box_transform ArgSpec(args=['input', 'name'], varargs=None, keywords=None, defaults=(None,))
+paddle.fluid.layers.yolov3_loss ArgSpec(args=['x', 'gtbox', 'gtlabel', 'anchors', 'class_num', 'ignore_thresh', 'loss_weight_xy', 'loss_weight_wh', 'loss_weight_conf_target', 'loss_weight_conf_notarget', 'loss_weight_class', 'name'], varargs=None, keywords=None, defaults=(None, None, None, None, None, None))
```
Comment (panyx0718) [inline paddle/fluid/API.spec:291]: Is everywhere calling this yolov3 loss? Doesn't seem to be a general name...

**Review by panyx0718 — DISMISSED** [review]

File: paddle/fluid/API.spec (hunk 11 above)
Comment (heavengate) [reply paddle/fluid/API.spec:291]: 论文作者自己写的框架是基于layer的，在yolo layer里面forward来计算loss，github上贡献者基于keras，pytorch实现的版本一般叫yolov3 loss，loss的计算与yolov2有一定区别，可能叫yolov3_loss相对清晰一些

**Review by qingqing01 — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]
LG API
